lib/core/include/irods/configuration_hook_manager.hpp

```diff
@@ -0,0 +1,43 @@
+#ifndef IRODS_CONFIGURATION_HOOK_MANAGER_HPP
+#define IRODS_CONFIGURATION_HOOK_MANAGER_HPP
+
+/// \file
+
+#include <nlohmann/json.hpp>
+
+#include <vector>
+#include <string>
+
+namespace irods::experimental::configuration_reload
+{
+    /// \brief Hooks are functions called with the pathname of the json value that has changed, as well as the new
+    /// operation performed, removal versus replacement versus insertion etc
+    ///
+    /// Hooks are invoked in two stages for each entry in a list of json patches
+    ///
+    /// The first is the "*CATCHALL*" stage, where hooks that are run for every
+    /// configuration change occurs. The second is based on the specific json
+    /// path of the changed configuration value.
+    ///
+    /// For example, if you change the delay server's sleep duration, the
+    /// "*CATCHALL*" hooks will be run first in order of addition, followed by
+    /// the hooks associated with
+    /// "/advanced_settings/delay_server_sleep_time_in_seconds", also in order
+    /// of addition.
+    using hook = std::function<void(const std::string&, const nlohmann::json&)>;
+
+    /// \brief Adds a hook to be run whenever a key on the list of hooks changes.
+    ///        \see hook
+    /// \param [in] hook_triggers An array or container of json paths to match.
```

Comment:
What are the requirements on the container? Support for begin/end iterators?

We need to add a statement stating the requirements on the container.